### ISSUE TITLE
Bug: visibility not coordinated with required fields

### ISSUE THREAD

**Issue by gevro** [susom/redcap-em-shazam#30]
Hi,
Shazam's visibility mirroring option has a bug with required fields. Fields that are hidden but require a value, are still being required a value even though they are hidden. This needs to be fixed.
Thanks.

**Comment by 123andy** [comment]
Thanks for the report.  What do you suggest?  Should I check before hiding a field and replicate the REDCap 'hide and delete' modal?  I guess that would be my first guess...

**Comment by gevro** [comment]
What should happen is: If a field is hidden, then the field is no longer required.
That is the behavior in normal REDCap without Shazam. So Shazam should replicate that behavior.

**Comment by 123andy** [comment]
Oh sorry - I mis read.. I thought you were talking about hiding existing data and not erasing it...  My mind just went on auto-complete... 

Will, you could mirror the shazam hiding logic on the field where you are having problems and that should take care of your issue, no?

**Comment by gevro** [comment]
No that doesn't work. Let me explain:
- REDCap normal behavior: a field can be specified as required, but then when branching logic hides it, it is no longer required.
- Shazam behavior: I am using Shazam mirror visibility option for the above field. Shazam correctly hides and shows the field based on the branching logic. However, even when the field is hidden, it is still being 'required'. This can be easily replicated when you submit the survey, it throws an error saying the hidden field is required even though it is hidden.

So basically, there is some bug in Shazam's mirror visibility feature that somehow doesn't incorporate the fact that required fields that are hidden should no longer be required.

**Comment by 123andy** [comment]
Hi gevro - sorry for being unclear.  I'm guessing the field you are using in data-shazam-mirror-visibility is NOT the same field that is being required, correct?
My suggestion is to copy whatever the 'branching logic' is in the shazam-mirror-visibility field to the required field.  This way, when shazam hides the field, REDCap will also know that the field is hidden and will not enforce the required field logic.

Make sense?
Here is an example:

![image](https://user-images.githubusercontent.com/7461553/73036195-a0af0180-3dff-11ea-90aa-02b56082d4ec.png)


```
<div class="container">
    <div class="row">
        <div class="col">
            <div class='shazam'>fruit:label</div>
            <div class='shazam'>fruit</div>
        </div>
        <div class="col" data-shazam-mirror-visibility="apple_type">
            <div class='shazam'>apple_type:label</div>
            <div class='shazam'>apple_type</div>
            <hr>
            <div class='shazam'>apple_color:label</div>
            <div class='shazam'>apple_color</div>
        </div>
    </div>    
</div>
```

So long as the logic used in the shazam mirror visibility field is copied to the branching logic of all fields contained inside html element, it should just work...  (unless I'm missing something)

**Comment by gevro** [comment]
You're right. The issue was that my fields were mirroring visibility of a different field than the one shazam was pointing to. Although that intuitively should have worked, it doesn't. But changing the mirroring visibility to point to the same field that shazam is pointing to fixed the problem.

**Comment by 123andy** [comment]
it is a heavy lift to re-do redcap javascript in shazam so i’d rather just leave this as a work around.   maybe i can add a FAQ to the documentation to cover this issue?   thanks so much for bringing it here to my attention.

-Andy

________________________________

**Comment by gevro** [comment]
Yeah, I think just an FAQ explanation would be fine.
